tests/test_components/test_autograd.py

```diff
@@ -116,39 +117,107 @@
 def use_emulated_run(monkeypatch):
     """If this fixture is used, the `tests.utils.run_emulated` function is used for simulation."""
 
+    import tidy3d
+
     if TEST_MODE in ("pipeline", "speed"):
-        import tidy3d.web.api.webapi as webapi
+        task_id_fwd = "task_fwd"
+        AUX_KEY_SIM_FIELDS_KEYS = "sim_fields_keys"
 
-        monkeypatch.setattr(webapi, "run", run_emulated)
-        _run_was_emulated[0] = True
+        cache = {}
 
-        # import here so it uses emulated run
-        from tidy3d.web.api.autograd import autograd
+        import tidy3d.web.api.webapi as webapi
 
-        reload(autograd)
+        # reload(tidy3d.web.api.autograd.autograd)
+        from tidy3d.web.api.autograd.autograd import (
+            AUX_KEY_SIM_DATA_FWD,
+            AUX_KEY_SIM_DATA_ORIGINAL,
+            postprocess_adj,
+            postprocess_fwd,
+        )
 
+        def emulated_run_fwd(simulation, task_name, **run_kwargs) -> td.SimulationData:
+            """What gets called instead of ``web/api/autograd/autograd.py::_run_tidy3d``."""
+            task_id_fwd = task_name
+            if run_kwargs.get("simulation_type") == "autograd_fwd":
+                sim_original = simulation
+                sim_fields_keys = run_kwargs["sim_fields_keys"]
+                # add gradient monitors and make combined simulation
+                sim_combined = sim_original.with_adjoint_monitors(sim_fields_keys)
+                sim_data_combined = run_emulated(sim_combined, task_name=task_name)
+
+                # store both original and fwd data aux_data
+                aux_data = {}
+
+                _ = postprocess_fwd(
+                    sim_data_combined=sim_data_combined,
+                    sim_original=sim_original,
+                    aux_data=aux_data,
+                )
+
+                # cache original and fwd data locally for test
+                cache[task_id_fwd] = copy.copy(aux_data)
+                cache[task_id_fwd][AUX_KEY_SIM_FIELDS_KEYS] = sim_fields_keys
+                # return original data only
+                return aux_data[AUX_KEY_SIM_DATA_ORIGINAL], task_id_fwd
+            else:
+                return run_emulated(simulation, task_name=task_name), task_id_fwd
+
+        def emulated_run_bwd(simulation, task_name, **run_kwargs) -> td.SimulationData:
+            """What gets called instead of ``web/api/autograd/autograd.py::_run_tidy3d_bwd``."""
+
+            task_id_fwd = task_name[:-8]
```

Comment:
Yeah I winced at that too. It's related to [this](https://github.com/flexcompute/tidy3d/blob/3a8c20972eb6eb3f19495b912f9a0d51d1bd62cb/tidy3d/web/api/autograd/autograd.py#L685-L692) which is also not that pretty. No obvious way to improve but I think fundamentally an internal rethink may be needed rather than specifically fix the cache grabbing in the test.